spectacles/validators.py

```diff
@@ -228,53 +231,93 @@ def validate(self, mode: str = "batch") -> List[SqlError]:
         )
-        tasks = []
+        task_calls: List[Tuple[Callable, Tuple]] = []
+        # first we gather all parameters for the tasks
         for model in self.project.models:
             for explore in model.explores:
-                if explore.dimensions:
-                    if mode == "batch" or (mode == "hybrid" and not explore.queried):
-                        logger.debug("Querying one explore at at time")
-                        task = loop.create_task(
-                            self._query_explore(session, model, explore)
-                        )
-                        tasks.append(task)
-                    elif mode == "single" or (mode == "hybrid" and explore.errored):
-                        logger.debug("Querying one dimension at at time")
-                        for dimension in explore.dimensions:
-                            task = loop.create_task(
-                                self._query_dimension(
-                                    session, model, explore, dimension
-                                )
+                if mode == "batch" or (mode == "hybrid" and not explore.queried):
+                    task_calls.append((self._query_explore, (session, model, explore)))
+                elif mode == "single" or (mode == "hybrid" and explore.errored):
+                    for dimension in explore.dimensions:
+                        task_calls.append(
+                            (
+                                self._query_dimension,
+                                (session, model, explore, dimension),
                             )
-                            tasks.append(task)
-
-        query_task_ids = list(loop.run_until_complete(asyncio.gather(*tasks)))
-        loop.run_until_complete(session.close())
-        loop.run_until_complete(asyncio.sleep(0.250))
-        loop.close()
-
-        MAX_QUERY_FETCH = 250
-
-        tasks_to_check = query_task_ids[:MAX_QUERY_FETCH]
-        del query_task_ids[:MAX_QUERY_FETCH]
-        logger.debug(f"{len(query_task_ids)} left in queue")
-        tasks_to_check, errors = self._get_query_results(tasks_to_check)
-
-        while tasks_to_check or query_task_ids:
-            number_of_tasks_to_add = MAX_QUERY_FETCH - len(tasks_to_check)
-            tasks_to_check.extend(query_task_ids[:number_of_tasks_to_add])
-            del query_task_ids[:number_of_tasks_to_add]
-            logger.debug(f"{len(query_task_ids)} left in queue")
-            tasks_to_check, more_errors = self._get_query_results(tasks_to_check)
-            errors.extend(more_errors)
-            if tasks_to_check or query_task_ids:
-                time.sleep(0.5)
+                        )
+
+        tasks: Set = set()
+        query_task_ids: List = []
+        errors = []
+        logger.debug(
+            "we are going to schedule %s tasks at concurrency of %s",
+            len(task_calls),
+            max_concurrency,
+        )
+        while task_calls:
```

Comment:
When I run this with `max_concurrency = 0`, I get stuck in this loop forever and queries are never launched. `query_task_ids` is empty as defined above, so it continually waits for non-existent query task IDs to complete. Any idea what's going on here?

I wonder if you need to change `if len(query_task_ids) >= max_concurrency:` to a `>`?